train.py

```diff
@@ -398,6 +399,9 @@ def loss_fn(pred, labels):
             optimizers.step()
             lr_schedulers.step()
 
+            if job_config.training.precompute_float8_dynamic_scale_for_fsdp:
+                precompute_float8_dynamic_scale_for_fsdp(model)
```

Comment:
maybe a noob question: could you briefly explain what this is doing?
I wonder since we are already using context functions for FP8, can we have a context and run it in a `.step()` function here, just like optimizer, lr scheduler, and profiler. This would make the code consistent.

Reply:
> could you briefly explain what this is doing

`precompute_float8_dynamic_scale_for_fsdp` is a for-loop over `model.parameters()`. it issues a single all-reduce for all parameters, ie `abs(max(param)) for param in model.parameters()` and save amax/scale as `param._precomputed_scale`. this speed up the training loop since we do not need to compute amax/scale for each parameters in the training loop

> we are already using context functions for FP8

do you refer to `set_enable_fsdp_fp8_all_gather` ? That's for model intiaitialization where we swap `nn.Linear` with user-defined float8 linear. `precompute_float8_dynamic_scale_for_fsdp` is for training loop

Reply:
per suggestion, raise error if `use_fp8_linear=False` or `enable_fsdp_fp8_all_gather =False`

Reply:
noob q: do we eventually want to just put this in fsdp2?

Reply:
It has to be done after optimizer step (since parameter values change). Are you suggesting to run this in the root module's pre-forward?

Reply:
Yeah anywhere between the n-1th optimizer step and the first all-gather in the nth step where fsdp2 has control (if there's any).

Reply:
That makes sense. I think one concern is that FSDP is agnostic to the fp8 all-gather. FSDP does not know that the `fsdp_pre_all_gather` and `fsdp_post_all_gather` of the `Float8Linear.weight`s are implemented to do fp8 all-gather, so at best, the user still would need to register a module forward pre-hook or something to run this method.

Reply:
Ah I see. Somehow I thought fsdp2 is fp8-aware